bee-signing/src/ternary/wots/mod.rs

```diff
@@ -0,0 +1,307 @@
+// Copyright 2020 IOTA Stiftung
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+// an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and limitations under the License.
+
+//! Winternitz One Time Signature scheme.
+//! https://eprint.iacr.org/2011/191.pdf.
+
+mod normalize;
+mod shake;
+mod sponge;
+
+pub use normalize::{normalize, Error as NormalizeError};
+pub use shake::{WotsShakePrivateKeyGenerator, WotsShakePrivateKeyGeneratorBuilder};
+pub use sponge::{WotsSpongePrivateKeyGenerator, WotsSpongePrivateKeyGeneratorBuilder};
+
+use crate::ternary::{PrivateKey, PublicKey, RecoverableSignature, Signature, SIGNATURE_FRAGMENT_LENGTH};
+
+use bee_common_derive::{SecretDebug, SecretDisplay, SecretDrop};
+use bee_crypto::ternary::{sponge::Sponge, HASH_LENGTH};
+use bee_ternary::{T1B1Buf, TritBuf, Trits, Tryte, T1B1};
+
+use thiserror::Error;
+use zeroize::Zeroize;
+
+use std::{
+    convert::TryFrom,
+    fmt::{self, Display, Formatter},
+    marker::PhantomData,
+};
+
+/// Errors occuring during WOTS operations.
+#[derive(Debug, Error, PartialEq)]
+pub enum Error {
+    /// Missing security level in generator.
+    #[error("Missing security level in generator.")]
+    MissingSecurityLevel,
+    /// Failed sponge operation.
+    #[error("Failed sponge operation.")]
+    FailedSpongeOperation,
+    /// Invalid entropy length.
+    #[error("Invalid entropy length, should be 243 trits.")]
```

Comment:
a nice to have would be to print the actual provided entropy length. 
```suggestion
    #[error("Invalid entropy length, should be 243 trits, but was {0}.")]
```
Same with the other errors down below.

Reply:
Done.